src/appleseed/renderer/modeling/aov/diagnosticaov.cpp

```diff
@@ -0,0 +1,501 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "diagnosticaov.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/aov/aovaccumulator.h"
+#include "renderer/kernel/aov/imagestack.h"
+#include "renderer/kernel/rendering/pixelcontext.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/kernel/shading/shadingresult.h"
+#include "renderer/modeling/aov/aov.h"
+#include "renderer/modeling/frame/frame.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/image/tile.h"
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Diagnostic AOV accumulator.
+    //
+
+    class DiagnosticAOVAccumulator
+      : public AOVAccumulator
+    {
+      public:
+        DiagnosticAOVAccumulator()
+        {
+        }
+    };
+
+    //
+    // Invalid Sample AOV accumulator.
+    //
+    //
+
+    const uint8 NoState = 0;
+    const uint8 InvalidSample = 1;
+    const uint8 CorrectSample = 2;
+
+    class InvalidSampleAOVAccumulator
+      : public AOVAccumulator
+    {
+      public:
+        explicit InvalidSampleAOVAccumulator(
+            Image& image)
+          : m_image(image)
+          , m_invalid_sample_count(0)
+        {
+        }
+
+        void on_tile_begin(
+                const Frame&                frame,
+                const size_t                tile_x,
+                const size_t                tile_y,
+                const size_t                max_spp) override
+        {
+            // Create a tile that stores samples hint.
+            const Tile& tile = frame.image().tile(tile_x, tile_y);
+
+            m_invalid_sample_tile.reset(
+                new Tile(tile.get_width(), tile.get_height(), 1, PixelFormatUInt8));
+            m_invalid_sample_tile->clear(Color<uint8, 1>(0));
+
+            const CanvasProperties& props = frame.image().properties();
+
+            m_tile_origin_x = static_cast<int>(tile_x * props.m_tile_width);
+            m_tile_origin_y = static_cast<int>(tile_y * props.m_tile_height);
+            m_tile_end_x = static_cast<int>(m_tile_origin_x + m_invalid_sample_tile->get_width() - 1);
```

Comment:
Traditionally "end" indices are exclusive, i.e. you don't subtract 1 here, and you change <= to < when checking if pixels fall inside the tile. Simpler math and increased consistency.